### ISSUE TITLE
[discussion] changes needed to integrate with resource interface 

### ISSUE THREAD

**Issue by dongahn** [flux-framework/flux-sched#662]
@milroy and @SteVwonder: 

As we discussed by phone, I am creating this ticket to flesh out some details for our task breakdown for our next 3-week sprint. The goal is to test fluxion alone with flux-core on `fluke` in a more meaningful way. Towards the goal, we need to modify fluxion to integrate it with flux-core's `resource` interface.

The immediate capability of `resource` is its support for compute nodes that join the system instance in a staggered fashion. Its `acquire` interface uses RPC streaming to respond to the remote caller (`sched-fluxion-resource` in our case) with the info on the newly joined nodes and their execution targets (in the `idset` format).

As I understood, this module will ultimately provide an interface that allows the remote callers to get those nodes that are going up and down as well.

As is, fluxion is still operational even with the introduction of `resource`. `sched-fluxion-resource` uses the blocking watch on per-rank hwloc xml file from KVS so its initialization will block until the very last node will join the system instance and dump its hwloc file. 

But to keep abreast of `resource` development and be meaningful for fluke testing, we need to put in place a baseline integration. 

### I. 
The least time consuming solution would be:
`sched-fluxion-resource` sets up a streaming RPC with `resource.acquire.` On each response, fluxion can go to the corresponding ranks in the idset and join the hwloc xml in KVS for each rank into the resource graph. I believe we already have all of the code that can do "unpack" each hwloc into the graph so the new code needed will be set up for  `resource.acquire` and move resource graph construction code into "runtime" (as opposed to "module init" time.)

### II. 
However, the above in and of itself won't be so future proof. As nodes are going up and down, you want to be able to mark a subgraph accordingly so that those unavailable resources will be excluded from scheduling at "runtime." This would probably entail *at least* the following:

1. We need a class member variable for [`resource_pool_t`](https://github.com/flux-framework/flux-sched/blob/master/resource/schema/resource_data.hpp#L41) to keep track of each resource's up/down state. (Let's decide whether we want the state to represent in this member just as "up" or "down" or we need to represent some other state.)
2. We then augment [the traverser's availability check](https://github.com/flux-framework/flux-sched/blob/master/resource/schema/resource_data.hpp#L47), not only based on the available resource count but also based on the up/down state.
3. We need an efficient way to fetch a subgraph corresponding to each execution target. We probably can model this after `by_path` kind index maps. 
There will be two strategies that we can consider here. A) Having this index map each rank to the entire subgraph consisting of all of the vertices and edges that rank owns; B) having this index map only the top level vertice(s) that the rank owns. IMHO, A) would be most conservative and probably leads to a correct implementation with respect to other more complex cases like a scheduler reload with different traversal types and such. Though B will be much faster, it doesn't feel like this is the one we want to trade off complexity with performance. It is on the critical path of the general scheduler performance. When the subgraph is updated, we need to make sure we update all of their metadata as well as the metadata of its ancestors. Since @milroy have already looked into this as part of his elastic job research, I thought @milroy is the right person to do this! 

Although path II is a lot more involved than I, my suggestion would be to go with II since that seems to be the most bang for the buck.

As we discussed, there will be a modification needed for `qmanager` and `sched-fluxion-resource` synchronization in addition:
- Need a way for `sched-fluxion-resource` to signal resource "join done" to `qmanager`: when all of the execution target joins, we need to expose this information to `qmanager` so that it will start to reject jobs based on "satisfiability". In general, this will require some more thoughts (what does "satisfiability" really mean when our scheduler can grow). But I propose we keep this as simple as possible for this resource of implementation.

I plan to tackle this task along with reworking my state reconstruction PR on top of the master.

Let me know if you have any questions on the above.

**Comment by milroy** [comment]
Thank you for this writeup, @dongahn.  I will ask additional questions here and loop in other people as needed.

**Comment by milroy** [comment]
Taking a step back for a moment, do we need the _up_/_down_ state to be encoded in Fluxion?  In broad strokes, I'm wondering if we can achieve all of the functionality you describe here by implementing subgraph add and update as well as delete in the Fluxion `resource_reader_hwloc`.

There would be challenges in this approach.  The first is that deleting a subgraph rather than marking it _down_ requires changing the Boost Adjacency List vertex container type (which in my testing requires modifications to other parts of Fluxion).

**Comment by milroy** [comment]
> I'm wondering if we can achieve all of the functionality you describe here 

More specifically, what I mean is that Fluxion can schedule as usual on any resources in the resource graph.  If a rank goes down, the corresponding subgraph can be removed.  That would require a `by_rank` map of some kind, though.

**Comment by dongahn** [comment]
> Taking a step back for a moment, do we need the up/down state to be encoded in Fluxion? In broad strokes, I'm wondering if we can achieve all of the functionality you describe here by implementing subgraph add and update as well as remove in the Fluxion resource_reader_hwloc.

I think we will need to track up/down. I will allow us to perform better satisfiability checks. As I see this, up/down and attach/detach are complimentary to one another.

**Comment by dongahn** [comment]
@milroy: we have 1PM one on one today so let's hash out some details there as well.

**Comment by milroy** [comment]
> In the current implementation a scheduler needs to export a sched.resource-status RPC which returns an R object for each of the "all", "allocated" and "down" resource sets.

\+

> I think we will need to track up/down. I will allow us to perform better satisfiability checks. As I see this, up/down and attach/detach are complimentary to one another.

Makes sense to me.  Thanks for the clarification @grondo and @dongahn.

**Comment by dongahn** [comment]
> When used with flux-core, the scheduler resources should initially be "down" until a notification from the core resource module marks them "up". That doesn't mean you can't initialize resources to the up state, then mark everything down before initiating the resource.acquire RPC though.

I think there would be a difference between sched-simple and fluxion. If the new scheme is that fluxion attaches a hwloc xml subgraph on each acquire response, I think "up" would be more appropriate? I would first go with "up" for the first PR and with the next PR to integrate this with `resource.acquire`, I would reassess the default value.

**Comment by dongahn** [comment]
>> Since we use JGF for state reconstruction, it seems that we need to emit the status.
> I may be wrong, but you may be able to do up/down state reconstruction just using the resource.acquire response. The resource module is tracking state for up/down/drained execution targets (ranks) and sends the current state along in the initial response.

@milroy: I think you are referring to the reconstruction of a job on fluxion reload.

I don't think you can reconstruct the availability state accurately only by looking at `RV1` of the job. The `RV1` will be sent to the job scheme when the job is allocated but a compute node that RV1 contains could be drained after this. 

It seems we will need a coordination between qmanager and `sched-fluxion-resource` for the full state reconstruction. I agree with @grondo. It appears we may not want to use the availability field of `RV1` to determine the final state of a resource. 

In this regard, I wonder if we even want to export the availability field to the `RV1`'s JGF.

**Comment by dongahn** [comment]
> If you are using jansson json_unpack() you can use s?i to indicate an optional key.

+1

**Comment by dongahn** [comment]
> In this regard, I wonder if we even want to export the availability field to the RV1's JGF.

@milroy: I read your original posting once again and I can see why having this will help you testing. You can create a JGF graph with some resources marked "down" and perform `match alloc` to see if those unavailable subgraph is excluded from scheduling. 

A middle ground would be:
The JGF writer only optionally writes this field when a resource is "down". If you do this, for `RV1` with flux-core, this field won't be exported: I don't see how a job will be allocated to the "down" resources. Yet, you can still export an JGF with the "down" status for your testing? The JGF reader will have to use jansson optional key support as well.

**Comment by dongahn** [comment]
>  If we use enum to track the status, there doesn't appear to be any option other than to use the s:i format for JSON packing (writer) and unpacking or (reader) the value. Is there a better approach?

`enum` and `s?:i` work for me. To ensure explicit enum field to int type safety, I would use C++ enum class and use explicit integer assignments to enum fields.  

Example: https://github.com/flux-framework/flux-sched/blob/master/qmanager/modules/qmanager_opts.hpp#L65

https://stackoverflow.com/questions/18335861/why-is-enum-class-preferred-over-plain-enum

**Comment by grondo** [comment]
>  I would first go with "up" for the first PR and with the next PR to integrate this with resource.acquire, I would reassess the default value.

That makes sense. Sorry, I didn't want to confuse anything in this PR, but just wanted to make the point that for Flux, all resources are "down" until the flux-core `resource` module says they are "up" by including the exec target ids in the `up` idset. (like you said, this only applies when using the `resource.acquire` interface, though)

> I agree with @grondo. It appears we may not want to use the availability field of RV1 to determine the final state of a resource.

I think you may be confusing _R_ and the response to the `resource.acquire` RPC that a Flux scheduler sends to the `resource` module. The `resource` module doesn't respond with an _R_ currently, but includes an idset of "up" and "down" resources. See the comment at the top of `acquire.c` [here](https://github.com/flux-framework/flux-core/blob/master/src/modules/resource/acquire.c#L15).

**Comment by dongahn** [comment]
> That makes sense. Sorry, I didn't want to confuse anything in this PR, but just wanted to make the point that for Flux, all resources are "down" until the flux-core resource module says they are "up" by including the exec target ids in the up idset. (like you said, this only applies when using the resource.acquire interface, though)

Understood.

So that we are on the same page and make sure I don't have a blindspot... The current plan is to fetch the rank-specific hwloc xml corresponding to the `idset`  from each `resource.acquire` response and to attach a subgraph into the resource graph. IOW, a lazy resource graph building. 

Does `sched-simple` build the entire resource model before `resource.acquire`?

**Comment by dongahn** [comment]
> I think you may be confusing R and the response to the resource.acquire RPC that a Flux scheduler sends to the resource module. The resource module doesn't respond with an R currently, but includes an idset of "up" and "down" resources. See the comment at the top of acquire.c here.

I think I understand the difference between each running job's R and resource info coming from `resource.acquire`. For now, Fluxion will more of less ignore `resource.acquire`'s resource info and only use idset.

Fluxion uses each job's `R` to reconstruct the scheduler state on module `reload`. So I was just pointing out we can't really reconstruct the status state using the job's R on module reload. In fact, I am proposing we don't even export the status field of the job's `R` to RV1. Instead we should rely on the status info from `resource.acquire.` 

Hope this clarifies.

**Comment by garlick** [comment]
One minor point:  the excluded resources are omitted from the initial resource response.
Those resources will also never be marked "up".

So what you're proposing seems like a good first step, but you might decide a job request is satisfiable when it is not if excluded resources are taken into account.

The way around this would be to parse the keys in the by_rank object and create an aggregate idset and only fetch the xml ranks that are present in the idset.

Or wait for Rv2.

**Comment by dongahn** [comment]
> The way around this would be to parse the keys in the by_rank object and create an aggregate idset and only fetch the xml ranks that are present in the idset.

Yeah, this is the approach that @milroy and I agree that we will do.

> So what you're proposing seems like a good first step, but you might decide a job request is satisfiable when it is not if excluded resources are taken into account.

Yap, this is our `qmanager`-`sched-fluxion-resource` synchronization task. We plan to implement the synchronization such that `qmanager` will get notified when all resources are acquired by `sched-fluxion-resource`. Only after that, a job will be rejected based on its satisifiability. 

> One minor point: the excluded resources are omitted from the initial resource response.
Those resources will also never be marked "up".

Sorry. I'm a bit confused. Could you clarify what is "the excluded resources" in this context? E.g., those compute nodes that have not joined yet?

**Comment by garlick** [comment]
Resources can be excluded from scheduling by configuration (currently by execution target).  Those execution targets do not appear in the idset keys of the R_lite returned in the first acquire response, and they do not appear in up/down idsets after that.

**Comment by dongahn** [comment]
@garlick: Thanks. It seems it would de best if the scheduler just doesn't see those excluded resources then. BTW, while I have you, how does the scheduler know when it acquired all of the resources of an instance for scheduling? I was imagining `acquire` will send the final response for this condition but I may just be hallucinating.

**Comment by garlick** [comment]
All resource are returned in the first response.  (In fluxion case, fetching the ranks in the `by_rank` keys and then you have them all).  That set should be used for satisfiability.

The set of resources that are up/available varies over time.

**Comment by milroy** [comment]
> A middle ground would be:
The JGF writer only optionally writes this field when a resource is "down". If you do this, for RV1 with flux-core, this field won't be exported: I don't see how a job will be allocated to the "down" resources. Yet, you can still export an JGF with the "down" status for your testing? The JGF reader will have to use jansson optional key support as well.

Upon further consideration I don't think I need the key to be output in the JGF writer.  I can make tests with `match allocate` and compare against the matched resources when a given subgraph is marked down.

**Comment by dongahn** [comment]
Here is my initial sketch for resource integration:

<img width="1309" alt="Screen Shot 2020-06-12 at 2 46 00 PM" src="https://user-images.githubusercontent.com/862123/84548847-75985200-acbb-11ea-9436-0672f00087b9.png">

**Comment by dongahn** [comment]
**sched-fluxion-resource.notify**:
  - Stream RPC to notify the caller for each scheduleable resource state change
  - the first response triggers qmanager to start its handshaking w/ job-manager (using schedutil)
  - a subsequence response represents a resource event that leads to the invocation of run_sched_loop  within qmanager
  
**grow()**:
   - grow the resource graph (i.e., attaching subgraphs) and the resulting resource graph becomes the basis for satisfiability
    
**shrink()**:
   - shrink the resource graph (i.e., detaching subgraphs) and the resulting resource graph becomes the basis for satisfiability 
   - semantics include: leaving the job(s) running on the removed subgraphs. 
   - TBD: need to understand whether we can cleanly handle "sched-fluxion-resource.cancel" issued by qmanager when it receives the corresponding FREE request.
   
**mark()**:
   - mark a resource subgraph up or down and this doesn’t affect satisfiability

**policy modification for backfilling per Issue #670**

**Comment by dongahn** [comment]
I talked w/ @milroy after my NGCE meeting and compared our notes. Our plan:

1. I will post an experimental PR (as early as today) to show my code that implements https://github.com/flux-framework/flux-sched/issues/662#issuecomment-643496117 above. This will not even compile because It will depend on `mark()` that will have to come from @milroy's PRs. But this will give @milroy a concrete idea about how his PRs will be used.

2. @milroy will tackle the implementation of `mark()` using the current 2 PRs and one additional PR.
- a new PR that introduces exec-target to subgraph map.
- the public interface within traverser class will change to `mark()` within PR #665  

We won't worry about `shrink()` for now. The only event from `resource.acquire` that maps to `shrink()` is from an administrative node exclusion. But this is currently reported as `down` so we can't distinguish `shrink()` vs. `mark()` anyway.

@milroy and I still need to work out some details as to how to efficiently remove a subgraph using BGL. We may end up needing to change the vertex container to be list instead of vector...  Overall, this is a part of our overall graph elasticity support and we need more time to get our implementation seamlessly.

**Comment by milroy** [comment]
Thanks for diagram and outline @dongahn.  Regarding subgraph removal, see the [BGL documentation](https://www.boost.org/doc/libs/1_73_0/libs/graph/doc/adjacency_list.html) section titled Iterator and Descriptor Stability/Invalidation.  Fluxion introduces additional considerations into vertex deletion, so we should discuss this in detail in the future.

**Comment by dongahn** [comment]
@milroy: I'm not ready to post an experimental PR (working some bugs). But the code that will need your piece:

```c++
 532     if (up && (rc = ctx->traverser->mark (up,
 533                                           resource_pool_t::status_t::up)) < 0) {
 534         flux_log_error (ctx->h, "%s: mark (up)", __FUNCTION__);
 535         goto done;
 536     }
 537     if (down
 538         && (rc = ctx->traverser->mark (down,
 539                                        resource_pool_t::status_t::down)) < 0) {
 540         flux_log_error (ctx->h, "%s: mark (down)", __FUNCTION__);
 541         goto done;
 542     }
```

**Comment by milroy** [comment]
Thanks @dongahn!

**Comment by dongahn** [comment]
@milroy: FYI -- I pushed my branch to my [repo](https://github.com/dongahn/flux-sched/blob/acquire2/resource/modules/resource_match.cpp#L552). There are a few tests that are failing, which I should take a look before posting this as an experimental PR but I thought this might be useful for your PRs. [Here](https://github.com/dongahn/flux-sched/blob/acquire2/resource/modules/resource_match.cpp#L552) is the major merge point with your PRs.

**Comment by milroy** [comment]
> grow the resource graph (i.e., attaching subgraphs) and the resulting resource graph becomes the basis for satisfiability

I forgot to ask: is `grow ()` in scope for PR #675?

**Comment by dongahn** [comment]
> I forgot to ask: is grow () in scope for PR #675?

grow has already been implemented since "unpack_at" was already implemented in hwloc reader.

**Comment by dongahn** [comment]
https://github.com/flux-framework/flux-sched/blob/1be7d51d2f05323e5df6caa87e2164eab5f890df/resource/modules/resource_match.cpp#L417

**Comment by milroy** [comment]
> grow has already been implemented since "unpack_at" was already there in hwloc reader.

I see, I was thinking of `grow` in the context of elasticity for subgraphs of size >= 1.  (Our recent research, in other words.)  Thanks for the clarification, I understand now!

**Comment by dongahn** [comment]
> I see, I was thinking of grow in the context of elasticity for subgraphs of size >= 1. (Our recent research, in other words.) Thanks for the clarification, I understand now!

It is the same thing :-)

**Comment by milroy** [comment]
Ok, I see that now.  I was hung up on the single vertex argument to `grow`, but then traced it to the `const vtx_t parent` in `walk_hwloc`.
